internal/controller/etcdopstask/reconcile_task_test.go

```diff
@@ -0,0 +1,496 @@
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var (
+				cl  client.Client
+				err error
+			)
+			if tc.task != nil {
+				if tc.patchFails {
+					patchErr := apierrors.NewInternalError(fmt.Errorf("patch failed"))
+					cl = testutils.NewTestClientBuilder().
+						WithScheme(kubernetes.Scheme).
+						WithObjects(tc.task).
+						RecordErrorForObjects(testutils.ClientMethodPatch, patchErr, client.ObjectKeyFromObject(tc.task)).
+						Build()
+				} else {
+					cl = setupFakeClient(tc.task, false)
+					err = cl.Create(context.TODO(), tc.task)
+					g.Expect(err).ToNot(HaveOccurred(), "Failed to create task")
+					if tc.ContainsFinalizer {
+						controllerutil.AddFinalizer(tc.task, druidapicommon.EtcdOpsTaskFinalizerName)
+						err = cl.Update(context.TODO(), tc.task)
+						g.Expect(err).ToNot(HaveOccurred(), "Failed to update task with finalizer")
+					}
+				}
+			} else {
+				cl = setupFakeClient(nil, false)
+			}
+
+			r := newTestReconciler(t, cl)
+
+			taskKey := client.ObjectKey{Name: "test-task", Namespace: "test-ns"}
+			if tc.task != nil {
+				taskKey = client.ObjectKeyFromObject(tc.task)
+			}
+			result := r.ensureTaskFinalizer(context.TODO(), r.logger, taskKey, nil)
+
+			if tc.expectError {
+				g.Expect(result.HasErrors()).To(BeTrue())
+				g.Expect(result.GetCombinedError()).To(HaveOccurred())
+				g.Expect(result.GetCombinedError().Error()).To(ContainSubstring(tc.errSubstring))
+				return
+			}
+
+			g.Expect(result).To(Equal(tc.expectedResult))
+			updatedTask := &druidv1alpha1.EtcdOpsTask{}
+			err = cl.Get(context.TODO(), taskKey, updatedTask)
+			g.Expect(err).ToNot(HaveOccurred())
+			g.Expect(controllerutil.ContainsFinalizer(updatedTask, druidapicommon.EtcdOpsTaskFinalizerName)).To(BeTrue())
+		})
+	}
+}
+
+// TestTransitionToPendingState tests the transitionToPendingState step function.
+func TestTransitionToPendingState(t *testing.T) {
+	g := NewGomegaWithT(t)
+	tests := []struct {
+		name           string
+		task           *druidv1alpha1.EtcdOpsTask
+		expectedResult ctrlutils.ReconcileStepResult
+	}{
+		{
+			name:           "Task state is not nil, i.e either Pending or InProgress. Skip the step.",
+			task:           newTestTask(ptr.To(druidv1alpha1.TaskStatePending)),
+			expectedResult: ctrlutils.ContinueReconcile(),
+		},
+		{
+			name:           "Task state is nil, should transition to Pending",
+			task:           newTestTask(nil),
+			expectedResult: ctrlutils.ContinueReconcile(),
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			cl := setupFakeClient(tc.task, true)
+			r := newTestReconciler(t, cl)
+			err := cl.Create(context.TODO(), tc.task)
+			g.Expect(err).To(BeNil())
+
+			result := r.transitionToPendingState(context.TODO(), r.logger, client.ObjectKeyFromObject(tc.task), nil)
+			updatedTask := &druidv1alpha1.EtcdOpsTask{}
+			err = cl.Get(context.TODO(), client.ObjectKeyFromObject(tc.task), updatedTask)
```

Comment:
Since PR 777 we have been using `context.Background`. Lets stick to one convention and over a period of time change everything that this. Since this is new code, lets already change that everywhere you are using `context.TODO` and replace it with `context.Background`